### PR TITLE
Fix GNU machine mode attribute support for CHERI

### PR DESCRIPTION
**Pull request by jrtc27**

### CODE REVIEW

File: clang/include/clang/Basic/TargetInfo.h [hunk 1]
```diff
@@ -768,10 +768,10 @@ class TargetInfo : public virtual TransferrableTargetInfo,
 
   /// Return the "preferred" register width on this target.
   virtual unsigned getRegisterWidth() const {
-    // Currently we assume the register width on the target matches the pointer
+    // Currently we assume the register width on the target matches the size
     // width, we can introduce a new variable for this if/when some target wants
     // it.
-    return PointerWidth;
+    return getTypeWidth(SizeType);
```
Comment (arichardson) [inline clang/include/clang/Basic/TargetInfo.h:774]: Possibly worth upstreaming?

File: clang/test/Sema/cheri/attr-mode.c [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+
+// The "pointer" machine mode is always valid on pointer types and yields the
+// same underlying type. NB: Unlike GCC, Clang does not allow the attribute to
+// be used on language-level pointers, only integer and floating-point types
+// (and, for CHERI, integer capability types).
+
+typedef __INTPTR_TYPE__ intptr_ptr __attribute__((mode(pointer)));
+typedef __UINTPTR_TYPE__ uintptr_ptr __attribute__((mode(pointer)));
+
+int intptr_ptr_test[__builtin_types_compatible_p(intptr_ptr, __INTPTR_TYPE__) ? 1 : -1];
```
Comment (arichardson) [inline clang/test/Sema/cheri/attr-mode.c:12]: Maybe use `_Static_assert` instead?

**Review by arichardson — APPROVED** [review]

File: clang/include/clang/Basic/TargetInfo.h (hunk 1 above)
Comment (jrtc27) [reply clang/include/clang/Basic/TargetInfo.h:774]: Not sure, depends if we end up changing our minds on what this should do, and it's not like we're going to get conflicts for this very often...

Comment (arichardson) [reply clang/include/clang/Basic/TargetInfo.h:774]: Yeah probably not worth the effort right now.

**Pull request merged**